[FEATURE] Make saving kubeconfig to state optional when creating a civo_kubernetes resource

### Description

When creating a civo_kubernetes resource the kubeconfig gets saved to state, this includes secret tokens that grant complete access to the cluster. This is potentially not very secure.

I can see many scenarios (perhaps most?) where a user may not want that at all. Users may want to just create a cluster and then fetch the configuration with the Civo CLI or Dashboard, etc.

## Acceptance Criteria

* Create a new boolean configuration input called `write_kubeconfig` or something like that. When `true` write kubeconfig to state, when `false` don't write it.
* Ideally it would be good to default this to false, but it may cause backward compatibility issues. I say we default it to false, but output a message to the user saying that if they want to keep it to add the `write_kubeconfig` flag and set it to `true`. For example:

`"Starting on version x kubeconfig is not written to state by default, if you wish to keep the kubeconfig configuration in state, please add the input write_kubeconfig and set it to true"`



### Screenshots

_No response_

### Additional information

_No response_

Thank you for reviewing it @uzaxirr 

For the sdk change part, actually I was getting the following error, every time I updated the configuration.

<img width="776" alt="Screenshot 2024-08-06 193933" src="https://github.com/user-attachments/assets/4802a250-85e4-48d7-99e0-09c8f6a8d6a9">

Debug log revealed that, we are passing an empty `config.` to `UpdateKubernetesCluster()` inside `resourceKubernetesClusterUpdate()`, which caused this error.

Had there been a field like `writeKubeconfig`, this possibly wouldn't have happened I believe. But this doesn't necessitate adding a field to `KubernetesClusterConfig` in the sdk, as you rightly pointed out.

<img width="777" alt="Screenshot 2024-08-06 195619" src="https://github.com/user-attachments/assets/2d3f7dfb-f062-4823-a772-cac5f2fc645f">


As a workaround to ensure we don't send an empty config, I did the following:

<img width="675" alt="Screenshot 2024-08-07 004019" src="https://github.com/user-attachments/assets/9232b89b-f7ec-474e-9f41-be74cb76a248">

I am not sure if this is the best way, but I couldn't think of another one. And this fixed the issue.

Furthermore in the `ValidateWriteKubeconfig` function, I extracted the provider version the user is using and compared it with the threshold version (the current `1.0.48`), the warning is shown if the provider's version is `<= 1.0.48`

```
func ValidateWriteKubeconfig(v interface{}, path cty.Path) diag.Diagnostics {
	var versionInfo versionInfo
	diags := diag.Diagnostics{}

	cmd := exec.Command("terraform", "version", "-json")
	output, err := cmd.Output()
	if err != nil {
		log.Printf("[ERROR] error running terraform show: %v\n", err)
		return diags
	}

	err = json.Unmarshal(output, &versionInfo)
	if err != nil {
		log.Printf("[ERROR] error parsing JSON: %v\n", err)
		return diags
	}
	versionField := "registry.terraform.io/civo/civo"
	currentProviderVersion := versionInfo.ProviderSelections[versionField]
	thresholdProviderVersion := "1.0.48"

	v1, err := version.NewSemver(currentProviderVersion)
	if err != nil {
		log.Println("[ERROR] error parsing the given version")
		return diags
	}
	v2, err := version.NewVersion(thresholdProviderVersion)
	if err != nil {
		log.Println("[ERROR] error parsing the given version")
		return diags
	}

	if v1.LessThanOrEqual(v2) {
		diags = append(diags, diag.Diagnostic{
			Severity: diag.Warning,
			Summary:  "Default kubeconfig behavior changed",
			Detail:   "Starting from version 1.0.49, kubeconfig will no longer be written to the Terraform state by default for the civo_kubernetes resource. This change is made to enhance security by preventing sensitive information from being stored in state files. If you want to retain kubeconfig in your state file, please update your configuration by adding the `write_kubeconfig` parameter and setting it to `true`. Example configuration: `write_kubeconfig = true`.",
		})
	}
	return diags
}
```
I have a question, is the error handling done right here? If other errors like `parsing error` occurs, I am giving user the benefit of doubt and not showing the warning. Is it the right approach to take?

<img width="908" alt="Screenshot 2024-08-06 232945" src="https://github.com/user-attachments/assets/a2e285f6-50ca-4050-85d1-6964e381f2be">

This is how the warning looks:

<img width="906" alt="Screenshot 2024-08-06 233136" src="https://github.com/user-attachments/assets/f3e43d6b-9410-40d4-aa21-dfc9c877c89f">


**Post changes, the creation/updation goes through as intended:**

<img width="287" alt="Screenshot 2024-08-07 001911" src="https://github.com/user-attachments/assets/819cf336-c9df-43b5-a2a4-c25a9577c860">

Kubeconfig hidden:

<img width="907" alt="Screenshot 2024-08-07 001305" src="https://github.com/user-attachments/assets/c90cd2ac-9da0-4d3e-a81b-fa33dcb5b93b">

Updating resource:

<img width="290" alt="Screenshot 2024-08-07 001856" src="https://github.com/user-attachments/assets/43e6669b-c092-452c-832b-475f8cfd6ac3">
<img width="341" alt="Screenshot 2024-08-07 001824" src="https://github.com/user-attachments/assets/825b9e3d-a7d2-41cc-ac75-e94baf83cb65">

Kubeconfig being shown in state:

<img width="909" alt="Screenshot 2024-08-07 001730" src="https://github.com/user-attachments/assets/74b7407c-fedd-4aa3-8998-34a087026abc">

Updating again:
<img width="293" alt="Screenshot 2024-08-07 001926" src="https://github.com/user-attachments/assets/5871a34a-4a7d-475a-a9fc-64a61c7f5caa">
<img width="608" alt="Screenshot 2024-08-07 002032" src="https://github.com/user-attachments/assets/fe42f36b-954f-4de0-aa94-b0b98aa3253f">

--- a/civo/kubernetes/resource_kubernetes_cluster.go
+++ b/civo/kubernetes/resource_kubernetes_cluster.go
@@ -129,7 +129,7 @@ func ResourceKubernetesCluster() *schema.Resource {
 				Optional:         true,
 				Default:          false,
 				Description:      "Whether to write the kubeconfig to state",
-				ValidateDiagFunc: utils.ValidateWriteKubeconfig,
+				ValidateDiagFunc: utils.ValidateProviderVersion,
 			},
 			"api_endpoint": {
 				Type:        schema.TypeString,

--- a/civo/kubernetes/resource_kubernetes_cluster.go
+++ b/civo/kubernetes/resource_kubernetes_cluster.go
@@ -124,6 +124,13 @@ func ResourceKubernetesCluster() *schema.Resource {
 				Sensitive:   true,
 				Description: "The kubeconfig of the cluster",
 			},
+			"write_kubeconfig": {
+				Type:             schema.TypeBool,
+				Optional:         true,
+				Default:          false,
+				Description:      "Whether to write the kubeconfig to state",
+				ValidateDiagFunc: utils.ValidateWriteKubeconfig,
+			},
 			"api_endpoint": {
 				Type:        schema.TypeString,
 				Computed:    true,
@@ -332,13 +339,20 @@ func resourceKubernetesClusterRead(_ context.Context, d *schema.ResourceData, m 
 	d.Set("tags", strings.Join(resp.Tags, " ")) // space separated tags
 	d.Set("status", resp.Status)
 	d.Set("ready", resp.Ready)
-	d.Set("kubeconfig", resp.KubeConfig)
+	// d.Set("kubeconfig", resp.KubeConfig)
 	d.Set("api_endpoint", resp.APIEndPoint)
 	d.Set("master_ip", resp.MasterIP)
 	d.Set("dns_entry", resp.DNSEntry)
 	// d.Set("built_at", resp.BuiltAt.UTC().String())
 	d.Set("created_at", resp.CreatedAt.UTC().String())
 	d.Set("firewall_id", resp.FirewallID)
+
+	writeKubeconfig := d.Get("write_kubeconfig").(bool)
+	if writeKubeconfig {
+		d.Set("kubeconfig", resp.KubeConfig)
+	} else {
+		d.Set("kubeconfig", "")
+	}
 
 	if err := d.Set("pools", flattenNodePool(resp)); err != nil {
 		return diag.Errorf("[ERR] error retrieving the pool for kubernetes cluster error: %#v", err)
@@ -422,6 +436,11 @@ func resourceKubernetesClusterUpdate(ctx context.Context, d *schema.ResourceData
 
 	if d.HasChange("tags") {
 		config.Tags = d.Get("tags").(string)
+	}
+
+	if d.HasChange("write_kubeconfig") {
+		// setting atleast one field inside the KubernetesClusterConfig, just to ensure we are not sending an empty config
+		config.FirewallID = d.Get("firewall_id").(string)
 	}
 
 	log.Printf("[INFO] updating the kubernetes cluster %s", d.Id())

--- a/go.mod
+++ b/go.mod
@@ -4,6 +4,7 @@ require (
 	github.com/civo/civogo v0.3.73
 	github.com/google/uuid v1.3.1
 	github.com/hashicorp/go-cty v1.4.1-0.20200414143053-d3edf31b6320
+	github.com/hashicorp/go-version v1.7.0
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.31.0
 	github.com/stretchr/testify v1.8.4
 	golang.org/x/crypto v0.25.0
@@ -30,7 +31,6 @@ require (
 	github.com/hashicorp/go-multierror v1.1.1 // indirect
 	github.com/hashicorp/go-plugin v1.6.0 // indirect
 	github.com/hashicorp/go-uuid v1.0.3 // indirect
-	github.com/hashicorp/go-version v1.6.0 // indirect
 	github.com/hashicorp/hc-install v0.6.2 // indirect
 	github.com/hashicorp/hcl/v2 v2.19.1 // indirect
 	github.com/hashicorp/logutils v1.0.0 // indirect

--- a/internal/utils/utils.go
+++ b/internal/utils/utils.go
@@ -6,17 +6,25 @@ package utils
 import (
 	"encoding/json"
 	"fmt"
-	"github.com/google/uuid"
+	"log"
+	"os/exec"
 	"regexp"
 	"sort"
 	"strconv"
 	"strings"
 	"sync"
 
+	"github.com/google/uuid"
+
 	"github.com/civo/civogo"
 	"github.com/hashicorp/go-cty/cty"
+	"github.com/hashicorp/go-version"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 )
+
+type versionInfo struct {
+	ProviderSelections map[string]string `json:"provider_selections"`
+}
 
 // ValidateName is a function to check if the name is valid
 func ValidateName(v interface{}, _ string) (ws []string, es []error) {
@@ -185,6 +193,48 @@ func ValidateClusterType(v interface{}, path cty.Path) diag.Diagnostics {
 			Severity: diag.Error,
 			Summary:  "Invalid Cluster Type",
 			Detail:   "The specified cluster type is invalid. Please choose either 'k3s' or 'talos'.",
+		})
+	}
+	return diags
+}
+
+// ValidateWriteKubeconfig function compares the provider version used by user with the threshold version of 1.0.48 and show warning accordingly
+func ValidateWriteKubeconfig(v interface{}, path cty.Path) diag.Diagnostics {
+	var versionInfo versionInfo
+	diags := diag.Diagnostics{}
+
+	cmd := exec.Command("terraform", "version", "-json")
+	output, err := cmd.Output()
+	if err != nil {
+		log.Printf("[ERROR] error running terraform show: %v\n", err)
+		return diags
+	}
+
+	err = json.Unmarshal(output, &versionInfo)
+	if err != nil {
+		log.Printf("[ERROR] error parsing JSON: %v\n", err)
+		return diags
+	}
+	versionField := "registry.terraform.io/civo/civo"
+	currentProviderVersion := versionInfo.ProviderSelections[versionField]
+	thresholdProviderVersion := "1.0.48"
+
+	v1, err := version.NewSemver(currentProviderVersion)
+	if err != nil {
+		log.Println("[ERROR] error parsing the given version")
+		return diags
+	}
+	v2, err := version.NewVersion(thresholdProviderVersion)
+	if err != nil {
+		log.Println("[ERROR] error parsing the given version")
+		return diags
+	}
+
+	if v1.LessThanOrEqual(v2) {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Warning,
+			Summary:  "Default kubeconfig behavior changed",
+			Detail:   "Starting from version 1.0.49, kubeconfig will no longer be written to the Terraform state by default for the civo_kubernetes resource. This change is made to enhance security by preventing sensitive information from being stored in state files. If you want to retain kubeconfig in your state file, please update your configuration by adding the `write_kubeconfig` parameter and setting it to `true`. Example configuration: `write_kubeconfig = true`.",
 		})
 	}
 	return diags

--- a/internal/utils/utils.go
+++ b/internal/utils/utils.go
@@ -22,7 +22,8 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 )
 
-type versionInfo struct {
+// VersionInfo stores Provider's version Info
+type VersionInfo struct {
 	ProviderSelections map[string]string `json:"provider_selections"`
 }
 
@@ -198,9 +199,9 @@ func ValidateClusterType(v interface{}, path cty.Path) diag.Diagnostics {
 	return diags
 }
 
-// ValidateWriteKubeconfig function compares the provider version used by user with the threshold version of 1.0.48 and show warning accordingly
-func ValidateWriteKubeconfig(v interface{}, path cty.Path) diag.Diagnostics {
-	var versionInfo versionInfo
+// ValidateProviderVersion function compares the current provider verson of the user with the threshold version and shows warning accordingly
+func ValidateProviderVersion(v interface{}, path cty.Path) diag.Diagnostics {
+	var versionInfo VersionInfo
 	diags := diag.Diagnostics{}
 
 	cmd := exec.Command("terraform", "version", "-json")
@@ -217,7 +218,7 @@ func ValidateWriteKubeconfig(v interface{}, path cty.Path) diag.Diagnostics {
 	}
 	versionField := "registry.terraform.io/civo/civo"
 	currentProviderVersion := versionInfo.ProviderSelections[versionField]
-	thresholdProviderVersion := "1.0.48"
+	thresholdProviderVersion := "1.0.49"
 
 	v1, err := version.NewSemver(currentProviderVersion)
 	if err != nil {
@@ -230,12 +231,27 @@ func ValidateWriteKubeconfig(v interface{}, path cty.Path) diag.Diagnostics {
 		return diags
 	}
 
+	lastStep := path[len(path)-1]
+	var field string
+	if step, ok := lastStep.(cty.GetAttrStep); ok {
+		field = step.Name
+	}
+
 	if v1.LessThanOrEqual(v2) {
-		diags = append(diags, diag.Diagnostic{
-			Severity: diag.Warning,
-			Summary:  "Default kubeconfig behavior changed",
-			Detail:   "Starting from version 1.0.49, kubeconfig will no longer be written to the Terraform state by default for the civo_kubernetes resource. This change is made to enhance security by preventing sensitive information from being stored in state files. If you want to retain kubeconfig in your state file, please update your configuration by adding the `write_kubeconfig` parameter and setting it to `true`. Example configuration: `write_kubeconfig = true`.",
-		})
+		if field == "write_password" {
+			diags = append(diags, diag.Diagnostic{
+				Severity: diag.Warning,
+				Summary:  "Default initial_password behavior changed",
+				Detail:   "Starting from version 1.0.50 the initial password is not written to state by default, if you wish to keep the initial password configuration in state, please add the input write_password and set it to true. Example configuration: `write_password = true`.",
+			})
+
+		} else if field == "write_kubeconfig" {
+			diags = append(diags, diag.Diagnostic{
+				Severity: diag.Warning,
+				Summary:  "Default kubeconfig behavior changed",
+				Detail:   "Starting from version 1.0.50, kubeconfig will no longer be written to the Terraform state by default for the civo_kubernetes resource. This change is made to enhance security by preventing sensitive information from being stored in state files. If you want to retain kubeconfig in your state file, please update your configuration by adding the `write_kubeconfig` parameter and setting it to `true`. Example configuration: `write_kubeconfig = true`.",
+			})
+		}
 	}
 	return diags
 }

